packages/web-pkg/src/composables/actions/spaces/useSpaceActionsDuplicate.ts

```diff
@@ -0,0 +1,144 @@
+  const ability = useAbility()
+  const clientService = useClientService()
+  const loadingService = useLoadingService()
+
+  const duplicateSpace = async (spaces: SpaceResource[]) => {
+    const existingSpace = spaces[0]
+    const projectSpaces: SpaceResource[] = store.getters['runtime/spaces/spaces'].filter(
+      (space: SpaceResource) => isProjectSpaceResource(space)
+    )
+    const newSpaceName = resolveFileNameDuplicate(existingSpace.name, '', projectSpaces)
+
+    try {
+      const { data: createdSpace } = await clientService.graphAuthenticated.drives.createDrive(
+        {
+          name: newSpaceName,
+          description: existingSpace.description
+        },
+        {}
+      )
+      let newSpace = buildSpace(createdSpace)
+
+      const existingSpaceFiles = await clientService.webdav.listFiles(existingSpace)
+
+      if (existingSpaceFiles.children.length) {
+        const queue = new PQueue({ concurrency: 4 })
+        const copyOps = []
+
+        for (const file of existingSpaceFiles.children) {
+          copyOps.push(
+            queue.add(() =>
+              clientService.webdav.copyFiles(existingSpace, file, newSpace, { path: file.name })
+            )
+          )
+        }
+        await Promise.all(copyOps)
+      }
+
+      if (existingSpace.spaceReadmeData || existingSpace.spaceImageData) {
+        const specialRequestData = {
+          special: []
+        }
+
+        if (existingSpace.spaceReadmeData) {
+          const newSpaceReadmeFile = await clientService.webdav.getFileInfo(newSpace, {
+            path: `.space/${existingSpace.spaceReadmeData.name}`
+          })
+          specialRequestData.special.push({
+            specialFolder: {
+              name: 'readme'
+            },
+            id: newSpaceReadmeFile.id
+          })
+        }
+
+        if (existingSpace.spaceImageData) {
+          const newSpaceImageFile = await clientService.webdav.getFileInfo(newSpace, {
+            path: `.space/${existingSpace.spaceImageData.name}`
+          })
+          specialRequestData.special.push({
+            specialFolder: {
+              name: 'image'
+            },
+            id: newSpaceImageFile.id
+          })
+        }
+
+        const { data: updatedDriveData } =
+          await clientService.graphAuthenticated.drives.updateDrive(
+            newSpace.id.toString(),
+            specialRequestData as Drive,
+            {}
+          )
+        newSpace = buildSpace(updatedDriveData)
+      }
+
+      store.commit('runtime/spaces/UPSERT_SPACE', newSpace)
+      store.dispatch('showMessage', {
+        title: $gettext('Space was duplicated successfully')
+      })
+    } catch (error) {
+      console.error(error)
+      store.dispatch('showErrorMessage', {
+        title: $gettext('Failed to duplicate space'),
+        error
+      })
+    }
+  }
+
+  const handler = ({ resources }: SpaceActionOptions) => {
+    return duplicateSpace(resources)
```

Comment:
Why not put the `duplicateSpace` body here inline? You can export the handler function with an alias as `duplicateSpace: handler` in line 142...

Reply:
i like it more this way somehow, but now we need it anyways